x/trigger/keeper/event_listener.go

```diff
@@ -0,0 +1,61 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	triggertypes "github.com/provenance-io/provenance/x/trigger/types"
+)
+
+// SetEventListener Adds the trigger to the event listener store.
+func (k Keeper) SetEventListener(ctx sdk.Context, trigger triggertypes.Trigger) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&trigger)
+	event, _ := trigger.GetTriggerEventI()
+	store.Set(triggertypes.GetEventListenerKey(event.GetEventPrefix(), trigger.GetId()), bz)
+}
+
+// RemoveEventListener Removes the trigger from the event listener store.
+func (k Keeper) RemoveEventListener(ctx sdk.Context, trigger triggertypes.Trigger) bool {
+	store := ctx.KVStore(k.storeKey)
+	event, _ := trigger.GetTriggerEventI()
+	key := triggertypes.GetEventListenerKey(event.GetEventPrefix(), trigger.GetId())
+	keyExists := store.Has(key)
+	if keyExists {
+		store.Delete(key)
+	}
+	return keyExists
+}
+
+// GetEventListener Gets the event listener from the store.
+func (k Keeper) GetEventListener(ctx sdk.Context, eventName string, triggerID triggertypes.TriggerID) (trigger triggertypes.Trigger, err error) {
+	store := ctx.KVStore(k.storeKey)
+	key := triggertypes.GetEventListenerKey(eventName, triggerID)
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return trigger, triggertypes.ErrEventNotFound
```

Comment:
I'm really not a fan of not found errors.
I much prefer using a reference for the return type so that `nil` can be returned.

It's much easier to check if `thing == nil` than `errors.Is(err, triggertypes.ErrEventNotFound)`. The main pain points are a) knowing that you should use `errors.Is` for that check, b) importing the correct "errors" package (there are lot of "errors" packages in various places, and when `errors.Is` is needed, you often have to have two of those imported at the same time), c) you always have to do some code digging to find out what error you need to check for. I.e. any time you go to use the func, you have to read through the code to know how best to handle what it returns.

Reply:
I'm open to either approach, and I know the Rewards module uses the NullObject pattern. So the trigger returned has an id of 0, which is why that id does not pass validation. I know other languages try to avoid null, but I'm not sure what the stance on it is in golang. I'm also not sure what our standard is for our codebase.

Reply:
I think most of our stuff returns `nil` instead of not-found errors, but not all of it. And there isn't really a standard even if you look at what the SDK does.

The main thing is that it's easier and safer to check if the returned thing is `nil` than it is to check if the returned error is a specific one. It gets further complicated when the sdk's errors are involved (like `triggertypes.ErrEventNotFound`) because the sdk's idea of wrapping an error doesn't actually wrap the error according to golang standards; it just creates a new one with the content of the old one. So you can't even use `errors.Is` reliably to do that check.

There's actually 3 patterns commonly used:
1. Return a concrete type and a boolean to indicate whether or not it was found.
2. Return a concrete type and error, and return a specific error when not found.
3. Return a pointer and error and return nil, nil when not found.

I guess there's a 4th pattern I've seen that I don't care for: Return just a pointer. If it's not found or can't be unmarshalled, return nil.

I don't like the way it was done in the rewards module, where it relies on `0` being an invalid value for a certain field, so you have to call the validate function to find out whether or not you actually got what you wanted.